src/jit/morph.cpp

```diff
@@ -6893,6 +6893,11 @@ bool Compiler::fgCanFastTailCall(GenTreeCall* callee)
     }
 #endif
 
+    assert(!callee->AreArgsComplete());
+
+    fgInitArgInfo(callee);
```

Comment:
It's worrying to me that `fgCanFastTailCall` now has side effects that are non-obvious.

For example, `fgMorphTailCall` has:
```
    assert(!fgCanFastTailCall(call));
```
and we should never call a function in an assert that can have side effects. Given that it already could print based on `COMPlus_JitReportFastTailCallDecisions` or dump into JitDump, it probably already shouldn't be in an assert.

`fgInitArgInfo` will only be run once, and this changes where it will be run for potential fast tail calls. It _looks_ from the code like that's ok, but it's very confusing and non-obvious.

Reply:
I was basically saying that it's quite strange to have an assert that calls a method that could generate output. Which is already the situation.

I like your idea about running fgCanFastTailCall computation once, and setting a flag for the result. Then just asserting on the result.

I was wondering if we could call fgInitArgInfo exactly once, in one place, and then fgCanFastTailCall can depend on (and assert on) it already being called. One way that perhaps that could happen is by refactoring fgMorphCall: The huge section of code at the top under `call->CanTailCall()` should be hoisted into a separate function, and we should get rid of the NO_TAIL_CALL goto. _Perhaps_ we could then call fgInitArgInfo before that `if (call->CanTailCall())`.

Reply:
> I was wondering if we could call fgInitArgInfo exactly once, in one place, and then fgCanFastTailCall can depend on (and assert on) it already being called.

I like that idea, though I think it's reasonable to defer it until after this is merged - what do you think?

Reply:
Yes, we can implement these changes incrementally.

Reply:
I agree with these comments and it should be done in a later PR.